lib/ts/nextjs.ts

```diff
@@ -127,6 +128,219 @@ export default class NextJS {
+        } catch (err) {
+            if (Session.Error.isErrorFromSuperTokens(err)) {
+                return {
+                    hasToken: err.type !== Session.Error.UNAUTHORISED,
+                    hasInvalidClaims: err.type === Session.Error.INVALID_CLAIMS,
+                    session: undefined,
+                    baseResponse,
+                    nextResponse: new Response("Authentication required", {
+                        status: err.type === Session.Error.INVALID_CLAIMS ? 403 : 401,
+                    }),
+                };
+            } else {
+                throw err;
+            }
+        }
+    }
+
+    static async getSSRSession(
+        cookies: Array<{ name: string; value: string }>,
+        headers: Headers,
+        options?: VerifySessionOptions,
+        userContext?: Record<string, any>
+    ): Promise<{
+        session: SessionContainer | undefined;
+        hasToken: boolean;
+        hasInvalidClaims: boolean;
+    }> {
+        let cookiesObj: Record<string, string> = Object.fromEntries(
+            cookies.map((cookie) => [cookie.name, cookie.value])
+        );
+
+        let baseRequest = new PreParsedRequest({
+            method: "get",
+            url: "",
+            query: {},
+            headers: headers,
+            cookies: cookiesObj,
+            getFormBody: async () => [],
+            getJSONBody: async () => [],
+        });
+
+        const { baseResponse, nextResponse, ...result } = await NextJS.commonSSRSession(
+            baseRequest,
+            options,
+            userContext
+        );
+        return result;
+    }
+
+    static async withSession<NextRequest extends PartialNextRequest, NextResponse extends Response>(
+        req: NextRequest,
+        handler: (error: Error | undefined, session: SessionContainer | undefined) => Promise<NextResponse>,
+        options?: VerifySessionOptions,
+        userContext?: Record<string, any>
+    ): Promise<NextResponse> {
+        try {
+            const query = Object.fromEntries(new URL(req.url).searchParams.entries());
+            const cookies: Record<string, string> = Object.fromEntries(
+                req.cookies.getAll().map((cookie) => [cookie.name, cookie.value])
+            );
+
+            let baseRequest = new PreParsedRequest({
+                method: req.method as HTTPMethod,
+                url: req.url,
+                query: query,
+                headers: req.headers,
+                cookies: cookies,
+                getFormBody: () => req!.formData(),
+                getJSONBody: () => req!.json(),
+            });
+
+            const { session, nextResponse, baseResponse } = await NextJS.commonSSRSession(
+                baseRequest,
+                options,
+                userContext
+            );
+
+            if (nextResponse) {
+                return nextResponse as NextResponse;
+            }
+
+            let userResponse = await handler(undefined, session);
```

Comment:
should wrap this with try / catch and call supertokens error handler in it before doing the above. Cause for example, if the user calls `session.mergeIntoAccessTokenPayload` in the `handler` func, and the session doesn't exist in the db, then we should throw a 401 UNAUTHORISED to the frontend, and that is done by the error handler

Reply:
Note that here, `userResponse` will be `undefined`, so you need to create your own nextresponse and copy into that the collectingResponse that's passed into the errorHandler

Reply:
done